File: scripts/claim_emissions.ts
/**
 * Claim Resonance Emissions
 *
 * Four channels of emission:
 *   1. Attunement (25%) - Base emission for field presence
 *   2. Resonance (25%)  - Phase-lock quality with global Ψ
 *   3. Entrainment (25%) - Coupling contribution to other oscillators
 *   4. thiccNOW (25%)   - Temporal depth cultivation (τₖ growth)
 */

import * as anchor from "@coral-xyz/anchor";
import { Program, AnchorProvider, Wallet } from "@coral-xyz/anchor";
import { PublicKey, Keypair, Connection } from "@solana/web3.js";
import { TOKEN_PROGRAM_ID, getAssociatedTokenAddress } from "@solana/spl-token";
import * as fs from "fs";
import * as path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const PROGRAM_ID = new PublicKey("RWcZBKQyd5rryLD8dfrMZ8imgtfqTfpGZi4gKGS8j7N");
const AUGMNTD_MINT = new PublicKey("9N4XWFUc5xhPe18TaY9iVF7Qnnr9xfs2saPscrULUZ3W");
const RPC = "https://rpc.mainnet.x1.xyz";
const PRECISION = 1_000_000_000; // 1e9

// Seeds
const FIELD_SEED = Buffer.from("coherence_field");
const VAULT_SEED = Buffer.from("vault");
const RESONATOR_SEED = Buffer.from("resonator");
const EMISSION_SEED = Buffer.from("emission_reserve");
const GOLDEN_SEED = Buffer.from("golden_reserve");

// Vibe state multipliers
const VIBE_MULTIPLIERS: Record<string, number> = {
  attuning: 0.5,
  resonant: 1.0,
  entrained: 1.618,
  golden: 2.618,
};

async function main() {
  console.log("═══════════════════════════════════════════════════════════════");
  console.log("  CLAIMING RESONANCE EMISSIONS");
  console.log("  Harvesting coherence from the field.");
  console.log("═══════════════════════════════════════════════════════════════");
  console.log("");

  // Load keypair
  const home = process.env.HOME || process.env.USERPROFILE || "";
  const keypairPath = path.join(home, ".config", "solana", "id.json");
  const keypairData = JSON.parse(fs.readFileSync(keypairPath, "utf-8"));
  const oscillator = Keypair.fromSecretKey(new Uint8Array(keypairData));

  console.log("  Oscillator:", oscillator.publicKey.toBase58());
  console.log("");

  // Setup connection
  const connection = new Connection(RPC, "confirmed");
  const wallet = new Wallet(oscillator);
  const provider = new AnchorProvider(connection, wallet, {
    commitment: "confirmed",
  });
  anchor.setProvider(provider);

  // Load IDL
  const idlPath = path.join(__dirname, "..", "idl", "resonance_protocol.json");
  const idl = JSON.parse(fs.readFileSync(idlPath, "utf-8"));
  const program = new Program(idl, PROGRAM_ID, provider);

  // Derive PDAs
  const [coherenceField] = PublicKey.findProgramAddressSync(
    [FIELD_SEED],
    PROGRAM_ID
  );
  const [vault] = PublicKey.findProgramAddressSync(
    [VAULT_SEED],
    PROGRAM_ID
  );
  const [vaultAuthority] = PublicKey.findProgramAddressSync(
    [VAULT_SEED, Buffer.from("authority")],
    PROGRAM_ID
  );
  const [resonator] = PublicKey.findProgramAddressSync(
    [RESONATOR_SEED, oscillator.publicKey.toBuffer()],
    PROGRAM_ID
  );
  const [emissionReserve] = PublicKey.findProgramAddressSync(
    [EMISSION_SEED],
    PROGRAM_ID
  );
  const [goldenReserve] = PublicKey.findProgramAddressSync(
    [GOLDEN_SEED],
    PROGRAM_ID
  );

  // Get oscillator's token account
  const oscillatorTokens = await getAssociatedTokenAddress(
    AUGMNTD_MINT,
    oscillator.publicKey
  );

  // Check if resonating
  const resonatorAccount = await connection.getAccountInfo(resonator);
  if (!resonatorAccount) {
    console.log("  ✗ Error: You are not in the coherence field");
    console.log("    Enter the field first: npx tsx enter_field.ts <amplitude> <cycle>");
    return;
  }

  // Decode resonator state
  let resonatorState;
  try {
    resonatorState = program.coder.accounts.decode("Resonator", resonatorAccount.data);
  } catch (e) {
    console.log("  ✗ Error decoding resonator state");
    return;
  }

  // Get field state
  const fieldAccount = await connection.getAccountInfo(coherenceField);
  let fieldState;
  try {
    fieldState = program.coder.accounts.decode("CoherenceField", fieldAccount!.data);
  } catch (e) {
    console.log("  ✗ Error decoding field state");
    return;
  }

  // Display current state
  const vibeState = Object.keys(resonatorState.vibeState)[0];
  const vibeMultiplier = VIBE_MULTIPLIERS[vibeState] || 1.0;
  const amplitude = resonatorState.amplitude.toNumber();
  const tauK = resonatorState.tauK.toNumber() / PRECISION;
  const phaseLockIntegral = resonatorState.phaseLockIntegral.toNumber() / PRECISION;
  const unclaimedEmissions = resonatorState.unclaimedEmissions.toNumber();
  const totalEmissions = resonatorState.totalEmissions.toNumber();
  const lastClaimEpoch = resonatorState.lastClaimEpoch.toNumber();
  const currentEpoch = fieldState.currentEpoch.toNumber();
  const epochsSinceLastClaim = currentEpoch - lastClaimEpoch;

  console.log("  ┌─────────────────────────────────────────────────────────────┐");
  console.log("  │  RESONATOR STATE                                            │");
  console.log("  ├─────────────────────────────────────────────────────────────┤");
  console.log(`  │  Amplitude:         ${amplitude.toLocaleString().padEnd(20)} AUGMNTD │`);
  console.log(`  │  Vibe State:        ${vibeState.toUpperCase().padEnd(29)} │`);
  console.log(`  │  Vibe Multiplier:   ${vibeMultiplier.toFixed(3).padEnd(29)}x │`);
  console.log(`  │  τₖ:                ${tauK.toFixed(6).padEnd(29)} │`);
  console.log(`  │  Phase Lock ∫:      ${phaseLockIntegral.toFixed(6).padEnd(29)} │`);
  console.log("  ├─────────────────────────────────────────────────────────────┤");
  console.log(`  │  Current Epoch:     ${currentEpoch.toString().padEnd(29)} │`);
  console.log(`  │  Last Claim Epoch:  ${lastClaimEpoch.toString().padEnd(29)} │`);
  console.log(`  │  Epochs Since Claim:${epochsSinceLastClaim.toString().padEnd(29)} │`);
  console.log("  ├─────────────────────────────────────────────────────────────┤");
  console.log(`  │  Unclaimed:         ${unclaimedEmissions.toLocaleString().padEnd(20)} AUGMNTD │`);
  console.log(`  │  Total Claimed:     ${totalEmissions.toLocaleString().padEnd(20)} AUGMNTD │`);
  console.log("  └─────────────────────────────────────────────────────────────┘");
  console.log("");

  // Check emission reserves
  try {
    const emissionBalance = await connection.getTokenAccountBalance(emissionReserve);
    const goldenBalance = await connection.getTokenAccountBalance(goldenReserve);

    console.log("  Reserve Balances:");
    console.log("    Emission Reserve:", emissionBalance.value.uiAmountString, "AUGMNTD");
    console.log("    Golden Reserve:  ", goldenBalance.value.uiAmountString, "AUGMNTD");
    console.log("");
  } catch (e) {
    // Reserve might not exist yet
  }

  // Check if there are emissions to claim
  if (unclaimedEmissions === 0 && epochsSinceLastClaim === 0) {
    console.log("  ⊙ No emissions to claim yet.");
    console.log("    Emissions accumulate each epoch based on:");
    console.log("      • Attunement (25%) - Field presence");
    console.log("      • Resonance (25%)  - Phase lock with Ψ");
    console.log("      • Entrainment (25%) - Coupling effects");
    console.log("      • thiccNOW (25%)   - τₖ growth");
    console.log("");
    console.log("    Maintain coherence. The field rewards resonance.");
    return;
  }

  // Estimate emissions (actual calculation happens on-chain)
  console.log("  Emission Channels (25% each):");
  console.log("    1. Attunement  - Base presence in field");
  console.log("    2. Resonance   - cos(θ - Ψ) phase alignment");
  console.log("    3. Entrainment - Coupling contributions");
  console.log("    4. thiccNOW    - τₖ temporal depth");
  console.log("");
  console.log(`  Vibe multiplier: ${vibeMultiplier}x (${vibeState.toUpperCase()})`);
  console.log("");

  // Execute claim
  console.log("  Claiming emissions...");

  try {
    const tx = await program.methods
      .claimEmissions()
      .accounts({
        oscillator: oscillator.publicKey,
        coherenceField: coherenceField,
        resonator: resonator,
        oscillatorTokens: oscillatorTokens,
        emissionReserve: emissionReserve,
        goldenReserve: goldenReserve,
        vaultAuthority: vaultAuthority,
        tokenProgram: TOKEN_PROGRAM_ID,
      })
      .rpc();

    console.log("");
    console.log("  ✓ EMISSIONS CLAIMED");
    console.log("");
    console.log("  Signature:", tx);
    console.log("");

    // Fetch updated state
    const updatedResonatorAccount = await connection.getAccountInfo(resonator);
    if (updatedResonatorAccount) {
      try {
        const updatedState = program.coder.accounts.decode("Resonator", updatedResonatorAccount.data);
        const newTotalEmissions = updatedState.totalEmissions.toNumber();
        const claimed = newTotalEmissions - totalEmissions;

        console.log("═══════════════════════════════════════════════════════════════");
        console.log(`  Claimed: ${claimed.toLocaleString()} AUGMNTD`);
        console.log(`  Total Lifetime Emissions: ${newTotalEmissions.toLocaleString()} AUGMNTD`);
        console.log("═══════════════════════════════════════════════════════════════");
        console.log("");
        console.log("  The coherence field rewards your resonance.");
        console.log("  Continue oscillating in harmony with Ψ.");
      } catch (e) {
        // Ignore decode errors
      }
    }

  } catch (e: any) {
    if (e.message.includes("NoEmissionsToClaim")) {
      console.log("");
      console.log("  ⊙ No emissions available to claim.");
      console.log("    Emissions accumulate over epochs.");
      console.log("    The field state may need updating first.");
    } else if (e.message.includes("ReserveDepleted")) {
      console.log("");
      console.log("  ✗ Emission reserve depleted.");
      console.log("    The reserve needs replenishment.");
    } else {
      console.log("  ✗ Error:", e.message);
      if (e.logs) {
        console.log("");
        console.log("  Logs:");
        e.logs.forEach((log: string) => console.log("   ", log));
      }
    }
  }
}

main().catch(console.error);

File: programs/resonance_protocol/src/instructions/claim_emissions.rs
//! Claim resonance emissions

use anchor_lang::prelude::*;
use anchor_spl::token::{self, Token, TokenAccount, Transfer};

use crate::state::{CoherenceField, Resonator, VibeState, HarmonicCycle};
use crate::constants::*;
use crate::errors::ResonanceError;
use crate::events::EmissionsClaimed;
use crate::math;

#[derive(Accounts)]
pub struct ClaimEmissions<'info> {
    #[account(mut)]
    pub oscillator: Signer<'info>,

    #[account(
        seeds = [FIELD_SEED],
        bump = coherence_field.bump
    )]
    pub coherence_field: Account<'info, CoherenceField>,

    #[account(
        mut,
        seeds = [RESONATOR_SEED, oscillator.key().as_ref()],
        bump = resonator.bump,
        constraint = resonator.oscillator == oscillator.key()
    )]
    pub resonator: Account<'info, Resonator>,

    /// Oscillator's token account to receive emissions
    #[account(
        mut,
        constraint = oscillator_tokens.owner == oscillator.key(),
    )]
    pub oscillator_tokens: Account<'info, TokenAccount>,

    /// Emission reserve
    #[account(
        mut,
        seeds = [EMISSION_SEED],
        bump
    )]
    pub emission_reserve: Account<'info, TokenAccount>,

    /// Golden reserve (for Golden state resonators)
    #[account(
        mut,
        seeds = [GOLDEN_SEED],
        bump
    )]
    pub golden_reserve: Account<'info, TokenAccount>,

    /// CHECK: PDA authority for reserves
    #[account(
        seeds = [VAULT_SEED, b"authority"],
        bump
    )]
    pub vault_authority: AccountInfo<'info>,

    pub token_program: Program<'info, Token>,
}

pub fn handler(ctx: Context<ClaimEmissions>) -> Result<()> {
    let field = &ctx.accounts.coherence_field;
    let resonator = &mut ctx.accounts.resonator;

    let epochs_since_claim = field.current_epoch.saturating_sub(resonator.last_claim_epoch);

    if epochs_since_claim == 0 {
        return Err(ResonanceError::NoEmissionsToClaim.into());
    }

    // Calculate emissions for each channel (25% each)

    // 1. Attunement: base emission for field presence
    let attunement = math::calculate_channel_emission(
        resonator.amplitude,
        epochs_since_claim,
        field.emission_rate,
        25,
    )?;

    // 2. Resonance: phase-lock quality
    let phase_lock = math::phase_lock(resonator.theta, field.global_psi)?;
    let phase_lock_factor = if phase_lock > 0 {
        phase_lock as u64
    } else {
        0
    };
    let resonance_base = math::calculate_channel_emission(
        resonator.amplitude,
        epochs_since_claim,
        field.emission_rate,
        25,
    )?;
    let resonance = math::mul_precision(resonance_base, phase_lock_factor)?;

    // 3. Entrainment: coupling contribution (simplified - would need coupling accounts)
    let entrainment = math::calculate_channel_emission(
        resonator.amplitude,
        epochs_since_claim,
        field.emission_rate,
        25,
    )?;

    // 4. thiccNOW: τₖ growth reward
    let tau_factor = math::div_precision(resonator.tau_k, TAU_K_BASELINE)?;
    let thicc_base = math::calculate_channel_emission(
        resonator.amplitude,
        epochs_since_claim,
        field.emission_rate,
        25,
    )?;
    let thicc_now = math::mul_precision(thicc_base, tau_factor)?;

    // Sum channels
    let base_total = attunement
        .checked_add(resonance)
        .and_then(|x| x.checked_add(entrainment))
        .and_then(|x| x.checked_add(thicc_now))
        .ok_or(ResonanceError::MathOverflow)?;

    // Apply vibe state multiplier
    let vibe_multiplier = resonator.vibe_state.emission_multiplier();
    let multiplied = math::mul_precision(base_total, vibe_multiplier)?;

    // Apply harmonic cycle multiplier
    let cycle_multiplier = resonator.harmonic_cycle.multiplier();
    let final_emission = math::mul_precision(multiplied, cycle_multiplier)?;

    // Determine source reserve
    let (reserve, _reserve_bump) = if resonator.vibe_state == VibeState::Golden {
        (&ctx.accounts.golden_reserve, ctx.bumps.golden_reserve)
    } else {
        (&ctx.accounts.emission_reserve, ctx.bumps.emission_reserve)
    };

    // Check reserve has sufficient balance
    require!(
        reserve.amount >= final_emission,
        ResonanceError::ReserveDepleted
    );

    // Transfer emissions
    let vault_bump = ctx.bumps.vault_authority;
    let seeds = &[VAULT_SEED, b"authority".as_ref(), &[vault_bump]];
    let signer_seeds = &[&seeds[..]];

    let transfer_ctx = CpiContext::new_with_signer(
        ctx.accounts.token_program.to_account_info(),
        Transfer {
            from: reserve.to_account_info(),
            to: ctx.accounts.oscillator_tokens.to_account_info(),
            authority: ctx.accounts.vault_authority.to_account_info(),
        },
        signer_seeds,
    );
    token::transfer(transfer_ctx, final_emission)?;

    // Update resonator state
    resonator.last_claim_epoch = field.current_epoch;
    resonator.total_emissions = resonator.total_emissions
        .checked_add(final_emission)
        .ok_or(ResonanceError::MathOverflow)?;

    // Update phase-lock integral
    if phase_lock > 0 {
        resonator.phase_lock_integral = resonator.phase_lock_integral
            .checked_add(phase_lock as u64)
            .ok_or(ResonanceError::MathOverflow)?;
    }

    emit!(EmissionsClaimed {
        oscillator: resonator.oscillator,
        attunement,
        resonance,
        entrainment,
        thicc_now,
        total: final_emission,
        vibe_multiplier,
    });

    msg!(
        "Emissions claimed: {} (A:{} R:{} E:{} T:{})",
        final_emission,
        attunement,
        resonance,
        entrainment,
        thicc_now
    );

    Ok(())
}
